sys/net/application_layer/cord/lc/cord_lc.c

```diff
@@ -0,0 +1,432 @@
+                       const sock_udp_ep_t *remote);
+/* callback for _send_rd_init_req() */
+static void _on_rd_init(const gcoap_request_memo_t *memo, coap_pkt_t *pdu,
+                        const sock_udp_ep_t *remote);
+static ssize_t _add_filters_to_lookup(coap_pkt_t *pkt, cord_lc_filter_t *filters);
+static int _send_rd_init_req(coap_pkt_t *pkt, const sock_udp_ep_t *remote,
+                             void *buf, size_t maxlen);
+/* do a lookup and returns payload of the response */
+static ssize_t _lookup_raw(const cord_lc_rd_t *rd, unsigned content_format,
+                           unsigned lookup_type, cord_lc_filter_t *filters,
+                           void *result, size_t maxlen);
+/* append default filters to requests, do raw lookup and decodes the raw response */
+ssize_t _lookup_result(cord_lc_rd_t *rd, cord_lc_res_t *result,
+                       cord_lc_filter_t *filters, void *buf, size_t maxlen,
+                       unsigned type);
+
+static char *_result_buf;
+static size_t _result_buf_len;
+
+static mutex_t _mutex = MUTEX_INIT;
+static volatile thread_t *_waiter;
+
+static void _lock(void)
+{
+    mutex_lock(&_mutex);
+    _waiter = sched_active_thread;
+}
+
+static int _sync(void)
+{
+    thread_flags_t flags = thread_flags_wait_any(FLAG_MASK);
+
+    if (flags & FLAG_ERR) {
+        return CORD_LC_ERR;
+    } else if (flags & FLAG_TIMEOUT) {
+        return CORD_LC_TIMEOUT;
+    } else if (flags & FLAG_OVERFLOW) {
+        return CORD_LC_OVERFLOW;
+    } else if (flags & FLAG_NORSC) {
+        return CORD_LC_NORSC;
+    } else {
+        return CORD_LC_OK;
+    }
+}
+
+static void _on_lookup(const gcoap_request_memo_t *memo, coap_pkt_t *pdu,
+                       const sock_udp_ep_t *remote)
+{
+    (void)remote;
+
+    thread_flags_t flag = FLAG_ERR;
+
+    if (memo->state == GCOAP_MEMO_RESP) {
+        unsigned ct = coap_get_content_type(pdu);
+        if (ct != COAP_FORMAT_LINK) {
+            DEBUG("cord_lc: unsupported content format\n");
+            goto end;
+        }
+        if (pdu->payload_len == 0) {
+            flag = FLAG_NORSC;
+            goto end;
+        }
+        if (pdu->payload_len >= _result_buf_len) {
+            flag = FLAG_OVERFLOW;
+            goto end;
+        }
+        memcpy(_result_buf, pdu->payload, pdu->payload_len);
+        memset(_result_buf + pdu->payload_len, 0,
+               _result_buf_len - pdu->payload_len);
+        _result_buf_len = pdu->payload_len;
+        flag = FLAG_SUCCESS;
+    } else if (memo->state == GCOAP_MEMO_TIMEOUT) {
+        flag = FLAG_TIMEOUT;
+    }
+
+end:
+    thread_flags_set((thread_t *)_waiter, flag);
+}
+
+static ssize_t _add_filters_to_lookup(coap_pkt_t *pkt, cord_lc_filter_t *filters)
+{
+    cord_lc_filter_t *f = filters;
+    while (f) {
+        for (unsigned i = 0; i < f->len; i++) {
+            char key[NANOCOAP_QS_MAX];
+            char val[NANOCOAP_QS_MAX];
```

Comment:
Allocating 128bytes per default (and possibly more if someone decides to redefine the value of `NANOCOAP_QS_MAX`) is not the best idea. Using a statically allocated buffer is IMO the better option here (see e.g. https://github.com/RIOT-OS/RIOT/pull/12774 for the problems this could cause)